File: job-scout/app/services/scrapers/providers/linkedin.py
import asyncio
import logging
from typing import List, Dict, Any, Optional
from datetime import datetime
import aiohttp
from bs4 import BeautifulSoup
import re
from urllib.parse import quote, urljoin
import time

from .base import BaseScraper, JobData
from app.db.models import JobSource

logger = logging.getLogger(__name__)


class LinkedInScraper(BaseScraper):
    """LinkedIn job scraper implementation using the proven API approach."""
    
    def __init__(self, search_settings: Dict[str, Any]):
        super().__init__(search_settings)
        self.base_url = "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search"
        self.session = None
    
    def get_job_source(self) -> JobSource:
        return JobSource.LINKEDIN
    
    def build_search_url(self, search_query: Dict[str, str]) -> str:
        """
        Build LinkedIn API search URL with parameters.
        
        Args:
            search_query: Dictionary with 'keywords', 'location', 'f_WT'
            
        Returns:
            Complete LinkedIn API search URL
        """
        keywords = quote(search_query.get('keywords', ''))
        location = search_query.get('location', '')
        f_wt = search_query.get('f_WT', '')
        timespan_code = self.search_settings.get('timespan_code', 'r84600')  # 24 hours default
        
        # Build URL parameters
        params = []
        if keywords:
            params.append(f"keywords={keywords}")
        if location:
            params.append(f"f_PP={location}")
        if f_wt:
            params.append(f"f_WT={f_wt}")
        
        # Add default parameters
        params.extend([
            "geoId=",
            f"f_TPR={timespan_code}",
            "start=0"
        ])
        
        url = f"{self.base_url}?{'&'.join(params)}"
        logger.info(f"Built LinkedIn API search URL: {url}")
        return url
    
    async def scrape_jobs(self, search_query: Dict[str, str]) -> List[JobData]:
        """
        Scrape jobs from LinkedIn based on search query using the API approach.
        
        Args:
            search_query: Dictionary with search parameters
            
        Returns:
            List of JobData objects
        """
        if not self.validate_search_query(search_query):
            raise ValueError("Invalid search query: missing required fields")
        
        jobs = []
        pages_to_scrape = self.search_settings.get('pages_to_scrape', 1)
        
        try:
            async with aiohttp.ClientSession() as session:
                self.session = session
                
                for page in range(pages_to_scrape):
                    logger.info(f"Scraping LinkedIn page {page + 1}/{pages_to_scrape}")
                    
                    # Build URL for this page
                    page_url = self.build_search_url(search_query)
                    page_url = page_url.replace("start=0", f"start={25 * page}")
                    
                    # Scrape this page
                    page_jobs = await self._scrape_page(page_url)
                    jobs.extend(page_jobs)
                    
                    # Exit if no results
                    if len(page_jobs) == 0:
                        break
                    
                    # Add delay between pages to be respectful
                    if page < pages_to_scrape - 1:
                        await asyncio.sleep(2)
        
        except Exception as e:
            logger.error(f"Error scraping LinkedIn jobs: {str(e)}")
            raise
        
        finally:
            self.session = None
        
        logger.info(f"Scraped {len(jobs)} jobs from LinkedIn")
        return jobs
    
    async def _scrape_page(self, url: str) -> List[JobData]:
        """
        Scrape a single page of job listings using LinkedIn API.
        
        Args:
            url: LinkedIn API search URL
            
        Returns:
            List of JobData objects from this page
        """
        jobs = []
        
        try:
            headers = {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36'
            }
            
            async with self.session.get(url, headers=headers, timeout=30) as response:
                if response.status != 200:
                    logger.error(f"Failed to fetch LinkedIn page: {response.status}")
                    return jobs
                
                html = await response.text()
                soup = BeautifulSoup(html, 'html.parser')
                
                # Parse job cards using the proven approach from old code
                jobs = self._transform_job_cards(soup)
        
        except Exception as e:
            logger.error(f"Error scraping LinkedIn page {url}: {str(e)}")
        
        return jobs
    
    def _transform_job_cards(self, soup: BeautifulSoup) -> List[JobData]:
        """
        Transform LinkedIn job cards into JobData objects using proven parsing logic.
        
        Args:
            soup: BeautifulSoup object of the page
            
        Returns:
            List of JobData objects
        """
        jobs = []
        
        try:
            divs = soup.find_all('div', class_='base-search-card__info')
        except:
            logger.info("Empty page, no jobs found")
            return jobs
        
        for item in divs:
            try:
                # Extract title
                title_elem = item.find('h3')
                title = title_elem.text.strip() if title_elem else "Unknown Title"
                
                # Extract company
                company_elem = item.find('a', class_='hidden-nested-link')
                company = company_elem.text.strip().replace('\n', ' ') if company_elem else "Unknown Company"
                
                # Extract location
                location_elem = item.find('span', class_='job-search-card__location')
                location = location_elem.text.strip() if location_elem else "Unknown Location"
                
                # Extract job URL from parent div
                parent_div = item.parent
                entity_urn = parent_div.get('data-entity-urn', '')
                job_posting_id = entity_urn.split(':')[-1] if entity_urn else ''
                job_url = f'https://www.linkedin.com/jobs/view/{job_posting_id}/' if job_posting_id else ''
                
                # Extract date
                date_tag_new = item.find('time', class_='job-search-card__listdate--new')
                date_tag = item.find('time', class_='job-search-card__listdate')
                date = None
                
                if date_tag and date_tag.get('datetime'):
                    try:
                        date = datetime.fromisoformat(date_tag['datetime'].replace('Z', '+00:00'))
                    except:
                        pass
                elif date_tag_new and date_tag_new.get('datetime'):
                    try:
                        date = datetime.fromisoformat(date_tag_new['datetime'].replace('Z', '+00:00'))
                    except:
                        pass
                
                # Create JobData object
                job_data = JobData(
                    title=title,
                    company=company,
                    location=location,
                    job_url=job_url,
                    date=date,
                    job_source=self.job_source
                )
                
                jobs.append(job_data)
                
            except Exception as e:
                logger.warning(f"Error parsing job card: {str(e)}")
                continue
        
        return jobs
    
    def parse_job_listing(self, job_element: Any) -> Optional[JobData]:
        """
        Parse a LinkedIn job listing element into JobData.
        This method is kept for compatibility but the main parsing is done in _transform_job_cards.
        
        Args:
            job_element: BeautifulSoup element representing a job card
            
        Returns:
            JobData object or None if parsing fails
        """
        # This method is not used in the new implementation
        # The main parsing is done in _transform_job_cards
        return None
    
    async def get_job_description(self, job_url: str) -> Optional[str]:
        """
        Get detailed job description from job URL using proven parsing logic.
        
        Args:
            job_url: URL of the specific job posting
            
        Returns:
            Job description text or None if failed
        """
        if not self.session:
            logger.error("No active session for fetching job description")
            return None
        
        try:
            headers = {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36'
            }
            
            async with self.session.get(job_url, headers=headers, timeout=30) as response:
                if response.status != 200:
                    return None
                
                html = await response.text()
                soup = BeautifulSoup(html, 'html.parser')
                
                # Use the proven description parsing logic from old code
                return self._transform_job_description(soup)
        
        except Exception as e:
            logger.error(f"Error fetching job description from {job_url}: {str(e)}")
            return None
    
    def _transform_job_description(self, soup: BeautifulSoup) -> str:
        """
        Transform job description using proven parsing logic from old code.
        
        Args:
            soup: BeautifulSoup object of the job page
            
        Returns:
            Cleaned job description text
        """
        div = soup.find('div', class_='description__text description__text--rich')
        if div:
            # Remove unwanted elements
            for element in div.find_all(['span', 'a']):
                element.decompose()

            # Replace bullet points
            for ul in div.find_all('ul'):
                for li in ul.find_all('li'):
                    li.insert(0, '-')

            text = div.get_text(separator='\n').strip()
            text = text.replace('\n\n', '')
            text = text.replace('::marker', '-')
            text = text.replace('-\n', '- ')
            text = text.replace('Show less', '').replace('Show more', '')
            return text
        else:
            return "Could not find Job Description"

    async def scrape_hardcoded_url(self, url_config: Dict[str, Any]) -> List[JobData]:
        """
        Scrape jobs from a hardcoded URL.
        
        Args:
            url_config: Dictionary with 'url', 'description', 'is_remote'
            
        Returns:
            List of JobData objects
        """
        url = url_config['url']
        is_remote = url_config.get('is_remote', False)
        pages_to_scrape = self.search_settings.get('pages_to_scrape', 1)
        
        jobs = []
        
        try:
            async with aiohttp.ClientSession() as session:
                self.session = session
                
                for page in range(pages_to_scrape):
                    logger.info(f"Scraping hardcoded URL page {page + 1}/{pages_to_scrape}")
                    
                    # Build URL for this page
                    page_url = url
                    if page > 0:
                        # Add pagination parameter
                        if 'start=' in page_url:
                            page_url = page_url.replace('start=0', f'start={25 * page}')
                        else:
                            page_url = f"{page_url}&start={25 * page}"
                    
                    # Scrape this page
                    page_jobs = await self._scrape_page(page_url)
                    
                    # Add remote flag to jobs if specified
                    if is_remote:
                        for job in page_jobs:
                            # You could add an is_remote field to JobData if needed
                            pass
                    
                    jobs.extend(page_jobs)
                    
                    # Exit if no results
                    if len(page_jobs) == 0:
                        break
                    
                    # Add delay between pages to be respectful
                    if page < pages_to_scrape - 1:
                        await asyncio.sleep(2)
        
        except Exception as e:
            logger.error(f"Error scraping hardcoded URL {url}: {str(e)}")
            raise
        
        finally:
            self.session = None
        
        logger.info(f"Scraped {len(jobs)} jobs from hardcoded URL: {url_config.get('description', url)}")
        return jobs
    
    @classmethod
    def get_default_settings(cls) -> Dict[str, Any]:
        """Get the default settings for LinkedIn scraper."""
        return {
            "job_source": JobSource.LINKEDIN,
            "search_queries": [
                {"keywords": "IT Help Desk", "location": "101076143", "f_WT": ""},
                {"keywords": "IT Help Desk", "location": "102252967", "f_WT": ""},
                {"keywords": "IT Help Desk", "location": "102570379", "f_WT": ""},
                {"keywords": "IT Help Desk", "location": "104198807", "f_WT": ""},
                {"keywords": "IT Help Desk", "location": "104779438", "f_WT": ""},
                {"keywords": "IT Help Desk", "location": "105135351", "f_WT": "3"},
                {"keywords": "IT Help Desk", "location": "105142029", "f_WT": "3"},
                {"keywords": "IT Help Desk", "location": "106362955", "f_WT": "3"},
                {"keywords": "IT Help Desk", "location": "105135351", "f_WT": "2"},
                {"keywords": "IT Help Desk", "location": "105142029", "f_WT": "2"},
                {"keywords": "IT Help Desk", "location": "106362955", "f_WT": "2"},
                {"keywords": "Application Support", "location": "101076143", "f_WT": ""},
                {"keywords": "Application Support", "location": "102252967", "f_WT": ""},
                {"keywords": "Application Support", "location": "102570379", "f_WT": ""},
                {"keywords": "Application Support", "location": "104198807", "f_WT": ""},
                {"keywords": "Application Support", "location": "104779438", "f_WT": ""},
                {"keywords": "Application Support", "location": "105135351", "f_WT": "3"},
                {"keywords": "Application Support", "location": "105142029", "f_WT": "3"},
                {"keywords": "Application Support", "location": "106362955", "f_WT": "3"},
                {"keywords": "Application Support", "location": "105135351", "f_WT": "2"},
                {"keywords": "Application Support", "location": "105142029", "f_WT": "2"},
                {"keywords": "Application Support", "location": "106362955", "f_WT": "2"}
            ],
            "hardcoded_urls": [
                {
                    "url": "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?keywords=IT%20Help%20Desk&location=United%20States&&f_WT=2&geoId=&f_TPR=r84600",
                    "description": "IT Help Desk - Remote - United States",
                    "is_remote": True
                },
                {
                    "url": "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?keywords=Application%20Support&location=United%20States&&f_WT=2&geoId=&f_TPR=r84600",
                    "description": "Application Support - Remote - United States",
                    "is_remote": True
                }
            ],
            "timespan_code": "r84600",
            "pages_to_scrape": 1,
            "rounds": 1
        }

File: job-scout/internal/scraper/linkedin.go
package scraper

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/jobscout/jobscout/internal/db"
	"golang.org/x/net/html"
)

const linkedInUserAgent = "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"

const linkedInBaseURL = "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search"

var linkedInPagePause = 5 * time.Second

// LinkedIn guest search 429s after enough paging; retry the same GET in-place
// instead of failing the whole scrape. Temporal does not retry the activity.
var (
	linkedInRetryAttempts  = 5
	linkedInRetryInitial   = 4 * time.Second
	linkedInRetryMaxWait   = 60 * time.Second
	linkedInRetryHeartbeat = 30 * time.Second
)

// linkedInWorkTypeOrder is the natural-language order for work-type phrases.
// LinkedIn ignores f_WT URL filters; selected types are prepended to keywords.
var linkedInWorkTypeOrder = []struct {
	code  string
	label string
}{
	{"2", "Remote"},
	{"3", "Hybrid"},
	{"1", "On-Site"},
}

// LinkedInScraper ports the original LinkedIn guest-API scraper to net/http +
// golang.org/x/net/html.
type LinkedInScraper struct {
	timespanCode  string
	pagesToScrape int
	client        *http.Client
	sleep         func(context.Context, time.Duration) error
}

func NewLinkedIn(settings db.ScraperSettings) *LinkedInScraper {
	return NewLinkedInWithTimeout(settings, 30*time.Second)
}

func NewLinkedInWithTimeout(settings db.ScraperSettings, timeout time.Duration) *LinkedInScraper {
	ts := settings.TimespanCode
	if ts == "" {
		ts = "r84600"
	}
	pages := settings.PagesToScrape
	if pages < 1 {
		pages = 1
	}
	if timeout <= 0 {
		timeout = 30 * time.Second
	}
	return &LinkedInScraper{
		timespanCode:  ts,
		pagesToScrape: pages,
		client:        &http.Client{Timeout: timeout},
	}
}

func (s *LinkedInScraper) Source() db.JobSource { return db.SourceLinkedIn }

// parseLinkedInWorkTypes returns selected work-type codes in phrase order.
// Empty or all three codes mean unrestricted (no keyword prefix).
func parseLinkedInWorkTypes(fWT string) []string {
	fWT = strings.TrimSpace(fWT)
	if fWT == "" {
		return nil
	}
	seen := map[string]bool{}
	for _, part := range strings.FieldsFunc(fWT, func(r rune) bool {
		return r == ',' || r == '|' || r == ' '
	}) {
		switch part {
		case "1", "2", "3":
			seen[part] = true
		}
	}
	if len(seen) == 0 || (seen["1"] && seen["2"] && seen["3"]) {
		return nil
	}
	out := make([]string, 0, len(seen))
	for _, wt := range linkedInWorkTypeOrder {
		if seen[wt.code] {
			out = append(out, wt.code)
		}
	}
	return out
}

func linkedInWorkTypePhrase(codes []string) string {
	if len(codes) == 0 {
		return ""
	}
	labels := make([]string, 0, len(codes))
	for _, code := range codes {
		for _, wt := range linkedInWorkTypeOrder {
			if wt.code == code {
				labels = append(labels, wt.label)
				break
			}
		}
	}
	switch len(labels) {
	case 0:
		return ""
	case 1:
		return labels[0]
	case 2:
		return labels[0] + " or " + labels[1]
	default:
		return strings.Join(labels[:len(labels)-1], ", ") + " or " + labels[len(labels)-1]
	}
}

func effectiveLinkedInKeywords(keywords, fWT string) string {
	phrase := linkedInWorkTypePhrase(parseLinkedInWorkTypes(fWT))
	if phrase == "" {
		return keywords
	}
	return phrase + " " + keywords
}

// combineLinkedInSearchQueries merges legacy one-code-per-row settings into one
// query per keywords+location with a comma-separated f_WT list.
func combineLinkedInSearchQueries(queries []map[string]string) []map[string]string {
	type key struct{ keywords, location string }
	order := make([]key, 0, len(queries))
	codesByKey := make(map[key]map[string]bool, len(queries))
	unrestricted := make(map[key]bool, len(queries))

	for _, q := range queries {
		k := key{keywords: q["keywords"], location: q["location"]}
		if _, ok := codesByKey[k]; !ok {
			order = append(order, k)
			codesByKey[k] = map[string]bool{}
		}
		codes := parseLinkedInWorkTypes(q["f_WT"])
		if codes == nil {
			unrestricted[k] = true
			continue
		}
		for _, code := range codes {
			codesByKey[k][code] = true
		}
	}

	out := make([]map[string]string, 0, len(order))
	for _, k := range order {
		fWT := ""
		if !unrestricted[k] {
			parts := make([]string, 0, 3)
			for _, wt := range linkedInWorkTypeOrder {
				if codesByKey[k][wt.code] {
					parts = append(parts, wt.code)
				}
			}
			if len(parts) > 0 && len(parts) < 3 {
				fWT = strings.Join(parts, ",")
			}
		}
		out = append(out, map[string]string{
			"keywords": k.keywords,
			"location": k.location,
			"f_WT":     fWT,
		})
	}
	return out
}

func (s *LinkedInScraper) buildSearchURL(query map[string]string, start int) string {
	params := []string{}
	if kw := query["keywords"]; kw != "" {
		params = append(params, "keywords="+url.QueryEscape(effectiveLinkedInKeywords(kw, query["f_WT"])))
	}
	if loc := query["location"]; loc != "" {
		params = append(params, "geoId="+loc)
	}
	// f_WT is settings-only; LinkedIn guest search ignores it as a URL filter.
	params = append(params, "f_TPR="+s.timespanCode, fmt.Sprintf("start=%d", start))
	return linkedInBaseURL + "?" + strings.Join(params, "&")
}

func (s *LinkedInScraper) ScrapeJobs(ctx context.Context, query map[string]string) ([]JobData, error) {
	if query["keywords"] == "" {
		return nil, fmt.Errorf("invalid search query: missing keywords")
	}

	var jobs []JobData
	start := 0
	firstCards := 0
	for page := 0; page < s.pagesToScrape; page++ {
		ReportProgress(ctx, Progress{
			Phase:         "page",
			Source:        string(db.SourceLinkedIn),
			Page:          page + 1,
			Pages:         s.pagesToScrape,
			Keywords:      query["keywords"],
			Location:      query["location"],
			JobsCollected: len(jobs),
			Message:       fmt.Sprintf("linkedin page %d/%d", page+1, s.pagesToScrape),
		})
		slog.Info("scraping linkedin page", "page", page+1, "of", s.pagesToScrape, "start", start)
		pageJobs, cards, err := s.scrapePage(ctx, s.buildSearchURL(query, start))
		if err != nil {
			return jobs, err
		}
		jobs = append(jobs, pageJobs...)
		if cards == 0 {
			break
		}
		start += cards
		if page == 0 {
			firstCards = cards
		} else if cards < firstCards {
			slog.Info("linkedin page short, stopping pagination",
				"page", page+1, "cards", cards, "first_page_cards", firstCards)
			break
		}
		if page < s.pagesToScrape-1 {
			if err := s.pause(ctx, linkedInPagePause); err != nil {
				return jobs, err
			}
		}
	}
	slog.Info("scraped linkedin jobs", "count", len(jobs))
	return jobs, nil
}

func (s *LinkedInScraper) scrapePage(ctx context.Context, pageURL string) ([]JobData, int, error) {
	resp, err := s.doGet(ctx, pageURL, "linkedin")
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()

	doc, err := html.Parse(resp.Body)
	if err != nil {
		return nil, 0, err
	}
	jobs, cards := s.transformJobCards(doc)
	return jobs, cards, nil
}

func (s *LinkedInScraper) transformJobCards(doc *html.Node) ([]JobData, int) {
	var jobs []JobData
	items := findAll(doc, "div", "base-search-card__info")
	for _, item := range items {
		title := text(findFirst(item, "h3", ""))
		if title == "" {
			title = "Unknown Title"
		}

		company := "Unknown Company"
		if c := findFirst(item, "a", "hidden-nested-link"); c != nil {
			company = strings.ReplaceAll(text(c), "\n", " ")
		}

		location := "Unknown Location"
		if l := findFirst(item, "span", "job-search-card__location"); l != nil {
			location = text(l)
		}

		jobURL := ""
		if item.Parent != nil {
			if urn := attr(item.Parent, "data-entity-urn"); urn != "" {
				parts := strings.Split(urn, ":")
				jobURL = "https://www.linkedin.com/jobs/view/" + parts[len(parts)-1] + "/"
			}
		}

		var date time.Time
		if t := findFirst(item, "time", ""); t != nil {
			if dt := attr(t, "datetime"); dt != "" {
				if parsed, err := time.Parse("2006-01-02", dt); err == nil {
					date = parsed
				}
			}
		}

		if jobURL == "" {
			continue // no stable URL means we cannot dedupe; skip
		}

		jobs = append(jobs, JobData{
			Title:    title,
			Company:  company,
			Location: location,
			JobURL:   jobURL,
			Date:     date,
			Source:   db.SourceLinkedIn,
		})
	}
	return jobs, len(items)
}

// ParseJobDescription extracts the posting body from LinkedIn job-detail HTML.
func ParseJobDescription(doc *html.Node) string {
	n := findFirst(doc, "div", "show-more-less-html__markup")
	return text(n)
}

// FetchJobDescription GETs job detail HTML and returns parsed text.
func (s *LinkedInScraper) FetchJobDescription(ctx context.Context, jobURL string) (string, error) {
	resp, err := s.doGet(ctx, jobURL, "linkedin detail")
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	doc, err := html.Parse(resp.Body)
	if err != nil {
		return "", err
	}
	return ParseJobDescription(doc), nil
}

func (s *LinkedInScraper) doGet(ctx context.Context, rawURL, errPrefix string) (*http.Response, error) {
	attempts := linkedInRetryAttempts
	if attempts < 1 {
		attempts = 1
	}
	backoff := linkedInRetryInitial
	for attempt := 1; attempt <= attempts; attempt++ {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("User-Agent", linkedInUserAgent)

		resp, err := s.client.Do(req)
		if err != nil {
			return nil, err
		}
		if resp.StatusCode == http.StatusOK {
			return resp, nil
		}

		status := resp.StatusCode
		header := resp.Header.Clone()
		_, _ = io.Copy(io.Discard, resp.Body)
		_ = resp.Body.Close()

		if !linkedInRetryable(status) || attempt == attempts {
			return nil, fmt.Errorf("%s returned status %d", errPrefix, status)
		}

		wait := linkedInRetryWait(header, backoff)
		slog.Warn("linkedin rate limited, retrying",
			"status", status,
			"attempt", attempt,
			"of", attempts,
			"wait", wait,
		)
		if err := s.pauseHeartbeat(ctx, wait, Progress{
			Phase:   "rate_limit",
			Source:  string(db.SourceLinkedIn),
			Message: fmt.Sprintf("linkedin %d, retry %d/%d in %s", status, attempt, attempts, wait),
		}); err != nil {
			return nil, err
		}
		if backoff > 0 {
			backoff *= 2
			if linkedInRetryMaxWait > 0 && backoff > linkedInRetryMaxWait {
				backoff = linkedInRetryMaxWait
			}
		}
	}
	return nil, fmt.Errorf("%s returned status 429", errPrefix)
}

func linkedInRetryable(status int) bool {
	return status == http.StatusTooManyRequests || status == http.StatusServiceUnavailable
}

func linkedInRetryWait(header http.Header, backoff time.Duration) time.Duration {
	wait := backoff
	raw := strings.TrimSpace(header.Get("Retry-After"))
	if raw != "" {
		if secs, err := strconv.Atoi(raw); err == nil && secs > 0 {
			wait = time.Duration(secs) * time.Second
		} else if at, err := http.ParseTime(raw); err == nil {
			if until := time.Until(at); until > 0 {
				wait = until
			}
		}
	}
	if linkedInRetryMaxWait > 0 && wait > linkedInRetryMaxWait {
		wait = linkedInRetryMaxWait
	}
	if wait < 0 {
		return 0
	}
	return wait
}

func (s *LinkedInScraper) pause(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return nil
	}
	if s.sleep != nil {
		return s.sleep(ctx, d)
	}
	return sleep(ctx, d)
}

func (s *LinkedInScraper) pauseHeartbeat(ctx context.Context, d time.Duration, progress Progress) error {
	if d <= 0 {
		return nil
	}
	ReportProgress(ctx, progress)
	chunk := linkedInRetryHeartbeat
	if chunk <= 0 {
		chunk = d
	}
	remaining := d
	for remaining > 0 {
		wait := remaining
		if wait > chunk {
			wait = chunk
		}
		if err := s.pause(ctx, wait); err != nil {
			return err
		}
		remaining -= wait
		if remaining > 0 {
			ReportProgress(ctx, progress)
		}
	}
	return nil
}

func sleep(ctx context.Context, d time.Duration) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(d):
		return nil
	}
}
